[IMP] mail,im_livechat: shortcut to partner form view

maad-odoo opened the pull request:
For Livechat Conversations:
Add a "View Contact" action that opens then partner record form view of the visitor, if the visitor has a partner record linked it it.

Enterprise PR: https://github.com/odoo/enterprise/pull/82059

task-4607618

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/203079.png)](https://mergebot.odoo.com/odoo/odoo/pull/203079)

maad-odoo commented on addons/mail/static/src/discuss/core/common/member_info_panel.js:
```diff
@@ -0,0 +1,96 @@
+import { useFileViewer } from "@web/core/file_viewer/file_viewer_hook";
+import { _t } from "@web/core/l10n/translation";
+
+export class CompactAttachmentList extends Component {
```
The current AttachmentList component displays more attachments than necessary in the sidebar info panel as this panel is meant to show more info the user and not the media that he had shared. 
For this purpose we need a compact version that shows only a few of the most recently uploaded attachments. Embedding this logic directly into the existing AttachmentList component would make it unnecessarily bulky. 

Hence, IMO, creating a separate, lightweight component for this compact view seems like a better approach :smile: .
the result, that we are looking for is:
![image](https://github.com/user-attachments/assets/58ef3c57-e4f3-49fa-b13a-7df0090de0d9)

Atovange replied:
But as the feature is implemented currently this is the only place to view attachments, which means the user is forced to use this reduced version of the attachment list.

maad-odoo replied:
Nope, the user won't be forced to only see the reduced version of the attachment list. Users will still have the option to view the Attachmentlist component as the "+2 more" is actually clickable, clicking on which will open the traditional Attachmentlist

![image](https://github.com/user-attachments/assets/ae8c166f-08e6-4890-b3d2-05915957b9d3)

maad-odoo replied:
The images can be accessed directly. Additionally, making the fileViewer to support other already supported file types is also possible (will be added) to show the preview. This will apply in cases where either of the first two attachments is not an image. Below is the UI for scenarios with two and one attachment:


[already added]
for two:
![image](https://github.com/user-attachments/assets/5ff5820f-a72d-40d1-bd96-0d35b230d6ea)
for one:
![image](https://github.com/user-attachments/assets/d8e8c02b-25dc-423b-b4e6-fbd43927d836)

Which case are you describing?

Atovange replied:
![image](https://github.com/user-attachments/assets/5c3e1695-c4a5-4628-9134-0a89245acb42)

In this case for instance I have no access to the existing AttachmentList.
Other than the display issues with the image, the js file is not clickable/downloadable and both attachments don't have the quick actions that attachments usually have.

By being at the bottom of the information panel I don't see the issue with displaying all the attachments with AttachmentList.

ThanhDodeurOdoo commented on addons/mail/static/src/discuss/core/common/thread_actions.js:
```diff
@@ -66,6 +68,9 @@ threadActionsRegistry
         sequenceGroup: 10,
         toggle: true,
+        hideInDiscuss(component) {
+            return component.env.inDiscussApp && component.thread.channel_type === "chat";
```
There are a few indirections so maybe there is something I'm missing, but t's a bit strange that the condition has `component.env.inDiscussApp` if the name of the predicate already suggests that it's in discuss.

maad-odoo replied:
Hii :wave: , the intention here is to not show the action in the threadActions header as I intend to show the compact version of the attachmentList first.

ThanhDodeurOdoo replied:
What I was wondering is if the component isn't always in discuss when this condition is checked (`component.env.inDiscussApp` would then be "always" true)

seb-odoo commented on addons/hr/models/res_partner.py:
```diff
@@ -62,3 +64,22 @@ def _compute_employee(self):
         for partner in self:
             partner.employee = partner in employees
+
+    def _to_store_defaults(self):
+
+        def get_employee_value(field_to_get):
+            def get_for_partner(partner):
+                # sudo: res.partner-fetch employees from the partner
+                matching_employees = partner.sudo().employee_ids.filtered(
+                    lambda employee: employee.company_id == partner.company_id
+                )
+                values = matching_employees.mapped(field_to_get) or [False]
+                return values[0]
+            return get_for_partner
+
+        fields = {
+            "department": get_employee_value("department_id.name"),
+            "designation": get_employee_value("job_title"),
+            "workplace": get_employee_value("work_location_name"),
+        }
+        return super()._to_store_defaults() + list(starmap(Store.Attr, fields.items()))
```
This should not be added to store in all flows, only the one that needs them.

seb-odoo replied:
Also we don't want to store employee value on partner, but introduce the employee model and make a relation between the two.

seb-odoo replied:
We might want to make a first PR without the HR overrides it is allows moving faster and refining this part later.

seb-odoo commented on addons/im_livechat/static/src/core/common/thread_actions.js:
```diff
@@ -0,0 +1,28 @@
+                search_view_id: [false],
+                res_id: visitor?.persona.id,
+            },
+            { newWindow: !!component.env.inDiscussApp }
```
Opening in new page doesn't feel great. It should always open in current window.
If discuss app is open, the current channel can be open as chat window at the same time as open the form view.

seb-odoo commented on addons/im_livechat/static/src/core/common/thread_actions.js:
```diff
@@ -0,0 +1,28 @@
+    },
+    open(component) {
+        const visitor = component.thread.livechatVisitorMember;
+        component.actionService?.doAction(
```
The button should probably not be shown if there is no `actionService`.

Is this in the wrong bundle? Maybe code was meant to be in `web` folder?

seb-odoo reviewed (commented):
Hello, thank you for the PR.

Could you focus on a simple version of the PR, where we only do the initial easy spec:
for livechat, if there is a partner, have the button to open the partner form.

The rest should be done separately.

Thank you!

maad-odoo commented:
@seb-odoo Hello, Thank you for your review :pray: 
I have cleaned the PR and kept only the code that is relevant to the original spec of the task. Also, about the demo-data, there is already session of Mitchel Admin and Joel Wills where this can be tested. Please let me know we need another one I will add it then.

seb-odoo commented on addons/mail/static/src/discuss/call/common/thread_actions.js:
```diff
@@ -24,11 +24,12 @@ threadActionsRegistry
         },
         sequence: 10,
-        sequenceQuick: 30,
+        sequenceQuick: (component) => !component.env.inDiscussApp && 30,
```
I'm not familiar with `sequenceQuick`, what is the goal of changing existing actions in the context of this PR?

seb-odoo commented:
There seems to be many issue on runbot, but none of them might be related to the PR... Could you rebase if possible to see if it's better?

maad-odoo commented:
@seb-odoo Hi 👋,
I've removed the actions definition from the registry and added a button in the side panel.
Here’s the updated UI:
![image](https://github.com/user-attachments/assets/7d3b29bc-d1b0-4737-9695-ac6ee550db9f)
Additionally, I’ve removed the logic for adding the open function to threadActionInternal and redefined the view-contact action specifically for WhatsApp channels.
Let me know if there’s anything else that needs to be addressed 😊

seb-odoo commented on addons/im_livechat/static/src/core/web/livechat_channel_info_list.xml:
```diff
@@ -2,6 +2,9 @@
 <templates xml:space="preserve">
     <t t-name="im_livechat.channelInfoList">
         <ActionPanel title.translate="Information" minWidth="200" initialWidth="250" icon="'fa fa-info'">
+            <button t-if="visitorPersona?.type === 'partner'" class="btn btn-primary mt-1" t-on-click="openVisitorContact" title="View Contact">
+                View Contact
+            </button>
```
As this is a link to a new page, it would be better as a real link with `a` and `href`, it can keep the custom on-click, but if the href is correct it also allows ctrl+click or middle+click to open the link in a new tab, or right click to copy the link.

seb-odoo commented on addons/im_livechat/static/src/core/web/livechat_channel_info_list.js:
```diff
@@ -22,4 +23,17 @@ export class LivechatChannelInfoList extends Component {
+    }
+
+    openVisitorContact() {
+        return this.actionService.doAction({
+            type: "ir.actions.act_window",
+            res_model: "res.partner",
+            views: [[false, "form"]],
+            res_id: this.visitorPersona?.id,
+        });
+    }
```
And to double check in small screen: probably don't open the chat window (otherwise the page is not visible), but maybe open a chat bubble?

See what do we when we open a crm link after /lead this is the same kind of idea.

maad-odoo commented on addons/im_livechat/demo/im_livechat_channel/im_livechat_session_15.xml:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0"?>
```
The existing demo session isn't directly accessible from Discuss — it has to be found through the sessions list.
As suggested by SEB in [this comment](https://github.com/odoo/odoo/pull/203079#issuecomment-2975913604), I’ve added a new demo record.

tsm-odoo reviewed (commented):
Thanks for the update. Almost there :)

tsm-odoo reviewed (approved):
@robodoo r+

tsm-odoo commented:
@robodoo r+
